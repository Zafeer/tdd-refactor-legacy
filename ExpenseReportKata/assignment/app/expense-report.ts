type ExpenseType = 'dinner' | 'breakfast' | 'car-rental';

type ExpenseDetails = {
  name: string;
  limit: number;
  mealCategory: boolean;
};

const ExpenseTypeDetails: Record<ExpenseType, ExpenseDetails> = {
  dinner: { name: 'Dinner', limit: 5000, mealCategory: true },
  breakfast: { name: 'Breakfast', limit: 1000, mealCategory: true },
  'car-rental': { name: 'Car Rental', limit: Infinity, mealCategory: false },
};

class Expense {
  type: ExpenseType;
  amount: number;
  constructor(type: ExpenseType, amount: number) {
    this.type = type;
    this.amount = amount;
  }
}

function getFormattedDate(): string {
  return new Date().toISOString().substr(0, 10);
}

interface ReportFormatter {
  generateHeader(): string;
  generateTableRow(expense: Expense): string;
  generateFooter(totalExpenses: number, mealExpenses: number): string;
}

class HtmlReportFormatter implements ReportFormatter {
  generateHeader(): string {
    return `<!DOCTYPE html>\n<html>\n<head>\n<title>Expense Report: ${getFormattedDate()}</title>\n</head>\n<body>\n<h1>Expense Report: ${getFormattedDate()}</h1>\n<table>\n<thead>\n<tr><th scope="col">Type</th><th scope="col">Amount</th><th scope="col">Over Limit</th></tr>\n</thead>\n<tbody>\n`;
  }

  generateTableRow(expense: Expense): string {
    const details = ExpenseTypeDetails[expense.type];
    const overLimitMarker = expense.amount > details.limit ? 'X' : ' ';
    return `<tr><td>${details.name}</td><td>${expense.amount}</td><td>${overLimitMarker}</td></tr>\n`;
  }

  generateFooter(totalExpenses: number, mealExpenses: number): string {
    return `</tbody>\n</table>\n<p>Meal Expenses: ${mealExpenses}</p>\n<p>Total Expenses: ${totalExpenses}</p>\n</body>\n</html>\n`;
  }
}

class PlainTextReportFormatter implements ReportFormatter {
  generateHeader(): string {
    return `Expense Report: ${getFormattedDate()}\n`;
  }

  generateTableRow(expense: Expense): string {
    const details = ExpenseTypeDetails[expense.type];
    const overLimitMarker = expense.amount > details.limit ? 'X' : ' ';
    return `${details.name}\t${expense.amount}\t${overLimitMarker}\n`;
  }

  generateFooter(totalExpenses: number, mealExpenses: number): string {
    return `Meal Expenses: ${mealExpenses}\nTotal Expenses: ${totalExpenses}\n`;
  }
}

function printReport(formatter: ReportFormatter, expenses: Expense[]): void {
  let totalExpenses = 0;
  let mealExpenses = 0;

  if (!formatter || typeof formatter.generateHeader !== 'function') {
    throw new TypeError(
      'Invalid formatter: must implement ReportFormatter interface'
    );
  }

  let output = formatter.generateHeader();

  for (const expense of expenses) {
    if (ExpenseTypeDetails[expense.type].mealCategory) {
      mealExpenses += expense.amount;
    }
    totalExpenses += expense.amount;
    output += formatter.generateTableRow(expense);
  }

  output += formatter.generateFooter(totalExpenses, mealExpenses);
  process.stdout.write(output);
}

export {
  printReport,
  Expense,
  ExpenseType,
  ReportFormatter,
  HtmlReportFormatter,
  PlainTextReportFormatter,
};
