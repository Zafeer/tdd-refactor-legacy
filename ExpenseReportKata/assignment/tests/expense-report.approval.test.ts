import { describe, expect, it, vi } from 'vitest';
import { printReport, Expense } from '../app/expense-report';

describe('Expense Report - Approval Tests', () => {
  describe('should generate a plain text report', () => {
    it('with no expenses', () => {
      //Arrange
      const mockStdoutWrite = vi
        .spyOn(process.stdout, 'write')
        .mockImplementation(() => true);
      const expenses: Expense[] = [];

      //Act
      printReport(false, expenses);
      const output = mockStdoutWrite.mock.calls.join('');
      const normalizedOutput = output.replace(/\d{4}-\d{2}-\d{2}/g, '{{DATE}}'); // Normalize the date (e.g., YYYY-MM-DD)

      // Assert
      expect(normalizedOutput).toMatchSnapshot();

      mockStdoutWrite.mockRestore();
    });

    it('with under and over limit expenses', () => {
      //Arrange
      const mockStdoutWrite = vi
        .spyOn(process.stdout, 'write')
        .mockImplementation(() => true);
      const expenses: Expense[] = [
        new Expense('dinner', 3000),
        new Expense('breakfast', 500),
        new Expense('car-rental', 10000),
        new Expense('dinner', 5001), // Over-limit dinner
        new Expense('breakfast', 1001), // Over-limit breakfast
        new Expense('car-rental', Number.MAX_SAFE_INTEGER), // Max car-rental
      ];

      //Act
      printReport(false, expenses);
      const output = mockStdoutWrite.mock.calls.join('');
      const normalizedOutput = output.replace(/\d{4}-\d{2}-\d{2}/g, '{{DATE}}'); // Normalize the date (e.g., YYYY-MM-DD)

      //Assert
      expect(normalizedOutput).toMatchSnapshot();

      mockStdoutWrite.mockRestore();
    });
  });

  describe('should generate a html report', () => {
    it('with no expenses', () => {
      //Arrange
      const mockStdoutWrite = vi
        .spyOn(process.stdout, 'write')
        .mockImplementation(() => true);
      const expenses: Expense[] = [];

      //Act
      printReport(true, expenses);
      const output = mockStdoutWrite.mock.calls.join('');
      const normalizedOutput = output.replace(/\d{4}-\d{2}-\d{2}/g, '{{DATE}}'); // Normalize the date (e.g., YYYY-MM-DD)

      // Assert
      expect(normalizedOutput).toMatchSnapshot();

      mockStdoutWrite.mockRestore();
    });
    it('with under and over limit expenses', () => {
      //Arrange
      const mockStdoutWrite = vi
        .spyOn(process.stdout, 'write')
        .mockImplementation(() => true);
      const expenses: Expense[] = [
        new Expense('dinner', 3000),
        new Expense('breakfast', 500),
        new Expense('car-rental', 10000),
        new Expense('dinner', 5001), // Over-limit dinner
        new Expense('breakfast', 1001), // Over-limit breakfast
        new Expense('car-rental', Number.MAX_SAFE_INTEGER), // Max car-rental
      ];

      //Act
      printReport(true, expenses);
      const output = mockStdoutWrite.mock.calls.join('');
      const normalizedOutput = output.replace(/\d{4}-\d{2}-\d{2}/g, '{{DATE}}'); // Normalize the date (e.g., YYYY-MM-DD)

      //Assert
      expect(normalizedOutput).toMatchSnapshot();

      mockStdoutWrite.mockRestore();
    });
  });
});
