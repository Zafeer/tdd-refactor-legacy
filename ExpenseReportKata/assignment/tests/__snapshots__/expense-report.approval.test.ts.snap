// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Expense Report - Approval Tests > should generate a html report > with no expenses 1`] = `
"<!DOCTYPE html>
<html>
<head>
<title>Expense Report: {{DATE}}</title>
</head>
<body>
<h1>Expense Report: {{DATE}}</h1>
<table>
<thead>
<tr><th scope="col">Type</th><th scope="col">Amount</th><th scope="col">Over Limit</th></tr>
</thead>
<tbody>
</tbody>
</table>
<p>Meal Expenses: 0</p>
<p>Total Expenses: 0</p>
</body>
</html>
"
`;

exports[`Expense Report - Approval Tests > should generate a html report > with under and over limit expenses 1`] = `
"<!DOCTYPE html>
<html>
<head>
<title>Expense Report: {{DATE}}</title>
</head>
<body>
<h1>Expense Report: {{DATE}}</h1>
<table>
<thead>
<tr><th scope="col">Type</th><th scope="col">Amount</th><th scope="col">Over Limit</th></tr>
</thead>
<tbody>
<tr><td>Dinner</td><td>3000</td><td> </td></tr>
<tr><td>Breakfast</td><td>500</td><td> </td></tr>
<tr><td>Car Rental</td><td>10000</td><td> </td></tr>
<tr><td>Dinner</td><td>5001</td><td>X</td></tr>
<tr><td>Breakfast</td><td>1001</td><td>X</td></tr>
<tr><td>Car Rental</td><td>9007199254740991</td><td> </td></tr>
</tbody>
</table>
<p>Meal Expenses: 9502</p>
<p>Total Expenses: 9007199254760492</p>
</body>
</html>
"
`;

exports[`Expense Report - Approval Tests > should generate a plain text report > with no expenses 1`] = `
"Expense Report: {{DATE}}
Meal Expenses: 0
Total Expenses: 0
"
`;

exports[`Expense Report - Approval Tests > should generate a plain text report > with under and over limit expenses 1`] = `
"Expense Report: {{DATE}}
Dinner	3000	 
Breakfast	500	 
Car Rental	10000	 
Dinner	5001	X
Breakfast	1001	X
Car Rental	9007199254740991	 
Meal Expenses: 9502
Total Expenses: 9007199254760492
"
`;
